oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java

```diff
@@ -0,0 +1,173 @@
+import static com.google.common.collect.Iterables.find;
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import java.util.List;
+import java.util.Set;
+import org.apache.jclouds.oneandone.rest.OneAndOneApi;
+import org.apache.jclouds.oneandone.rest.domain.DataCenter;
+import org.apache.jclouds.oneandone.rest.domain.Hdd;
+import org.apache.jclouds.oneandone.rest.domain.Server;
+import org.apache.jclouds.oneandone.rest.domain.ServerIp;
+import org.apache.jclouds.oneandone.rest.domain.Types.OSFamliyType;
+import org.apache.jclouds.oneandone.rest.domain.Types.ServerState;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.HardwareBuilder;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.compute.domain.NodeMetadataBuilder;
+import org.jclouds.compute.domain.OperatingSystem;
+import org.jclouds.compute.domain.OsFamily;
+import org.jclouds.compute.domain.Processor;
+import org.jclouds.compute.domain.Volume;
+import org.jclouds.compute.functions.GroupNamingConvention;
+import org.jclouds.domain.Location;
+import static org.jclouds.location.predicates.LocationPredicates.idEquals;
+import org.jclouds.util.InetAddresses2;
+
+public class ServerToNodeMetadata implements Function<Server, NodeMetadata> {
+
+   private final Supplier<Set<? extends Location>> locations;
+   private final Function<Hdd, Volume> fnVolume;
+   private final Function<List<ServerIp>, List<String>> fnCollectIps;
+   private final OneAndOneApi api;
+
+   private final GroupNamingConvention groupNamingConvention;
+
+   @Inject
+   public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
+           @Memoized Supplier<Set<? extends Location>> locations,
+           OneAndOneApi api,
+           GroupNamingConvention.Factory groupNamingConvention) {
+      this.locations = locations;
+      this.api = api;
+      this.fnVolume = fnVolume;
+      this.fnCollectIps = new Function<List<ServerIp>, List<String>>() {
+         @Override
+         public List<String> apply(List<ServerIp> in) {
+            List<String> ips = Lists.newArrayListWithExpectedSize(in.size());
+            for (ServerIp ip : in) {
+               ips.add(ip.ip());
+            }
+            return ips;
+         }
+      };
+      this.groupNamingConvention = groupNamingConvention.createWithoutPrefix();
+   }
+
+   @Override
+   public NodeMetadata apply(final Server server) {
+      checkNotNull(server, "Null server");
+
+      DataCenter dataCenter = api.dataCenterApi().get(server.datacenter().id());
+      Location location = find(locations.get(), idEquals(dataCenter.location()));
+
+      double size = 0d;
+      List<Volume> volumes = Lists.newArrayList();
+      List<Hdd> storages = server.hardware().hdds();
+      if (storages != null) {
+         for (Hdd storage : storages) {
+            size += storage.size();
+            volumes.add(fnVolume.apply(storage));
+         }
+      }

```

Comment:
Always building a custom hardware object? This seems not right. You should take into account prefedined flavours.